Improved ExtractionTest Validation of Attachment and Extracted Children Expected vs Found Count

sambish5 opened the pull request:
This is in reference to NationalSecurityAgency/burrito-grande#884.
Improved the ExtractionTest to further checks on Attachment and Extracted Children counts. Includes new tests in TestExtractionTest to verify count check works.

dev-mlb commented on src/main/java/emissary/test/core/ExtractionTest.java:
```diff
@@ -155,10 +153,44 @@ protected void checkAnswers(Document answers, IBaseDataObject payload, List<IBas
+        // check to see if payload has extracted records and see if attachments is not null
+        // if true, implement check on payload and attachments to compare expected v found counts
+        // else, implement checkAttachmentAndExtractedRecordCount
+        if ((attachments != null) || payload.hasExtractedRecords()) {
```
One case that does not work is if the <numAttachments>0</numAttachments> and you add <att1>...</att1>, this test still passes.

sambish5 replied:
Fixed the issue in the newest commit. Test now fails when numAttachments=0 and if att1 (or more) is present.

dev-mlb replied:
Same issue with extracted records.

dev-mlb replied:
If I don't define numAttachments and/or extractCount, tests pass.

dev-mlb commented on src/main/java/emissary/test/core/ExtractionTest.java:
```diff
@@ -256,7 +257,27 @@ protected void checkAnswers(Element el, IBaseDataObject payload, List<IBaseDataO
 
+    public void checkAttachmentAndExtractedRecordCount(Element el, IBaseDataObject payload, List<IBaseDataObject> attachments, String tname)
+            throws DataConversionException {
+        int numAtt = JDOMUtil.getChildIntValue(el, "numAttachments");
```
If I don't define numAttachments, then I can still have att elements in the xml and the test do not fail.

sambish5 replied:
Yep, after I stopped working last night I thought of a lot of these cases. Will push a new commit today to look out for these use cases.

dev-mlb commented on src/main/java/emissary/test/core/ExtractionTest.java:
```diff
@@ -256,7 +257,27 @@ protected void checkAnswers(Element el, IBaseDataObject payload, List<IBaseDataO
+        int numAtt = JDOMUtil.getChildIntValue(el, "numAttachments");
+        int attFoundCount = 0;
+        if (numAtt > -1) {
+            int attCount = 1;
```
You should not assume that the att count starts at one. Test will pass if you add att2 element.

dev-mlb commented on src/main/java/emissary/test/core/ExtractionTest.java:
```diff
@@ -1,8 +1,6 @@
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertTrue;
-import static org.junit.Assert.fail;
+import static org.junit.Assert.*;
```
Fix import

dev-mlb reviewed (changes requested):
lgtm - just need to fix the import statements

dev-mlb commented on src/test/java/emissary/test/core/TestExtractionTest.java:
```diff
@@ -1,5 +1,7 @@
 package emissary.test.core;
 
+import static org.junit.Assert.*;
```
import here too

drivenflywheel commented:
This seems way more burdensome than the original ask, which was "don't allow the XML to have more child attachment or extractedRecord elements than specified in the `numAttachments` / `extractionCount` elements". 

As currently implemented, we can no longer assert the numAttachments or extractionCount without creating a child element for every one of those children.  That could cause an enormous burden for downstream integration. and it will actively hinder extraction tests for some packaged file processing. If nothing else, the ExtractionTest class should provide the ability to disable this check on a class-by-class or test-by-test basis so downstream projects can integrate this version of emissary without requiring a huge integration PR with thousands of lines of new XML.

sambish5 commented:
Some updates:
- Added all changes into junit5/ExtractionTest file to pair with core/ExtractionTest
- Added check for answer file child `skipAttExtCountCheck`, which if set to true, allows validation to be skipped.

sambish5 commented:
@drivenflywheel @dev-mlb  some major updates within this. Instead of having it fail if `<numAttachments>` and `<extractCount>` do not match `<att#>` and `<extract#>`, it will now just log the non-matching counts. I currently have it written so even if numAttachments/extractCount are not specified, the check still occurs (this can easily be moved to within the `if > -1` parts if we only want it to check when those are specified, but I figured we'd want it to always be specified if there are att or extract in the xml).

If you guys would prefer that method however, I can move it inside of those ifs, and in the future if we want to build back out the functionality to fail if the counts do not match, we can.

drivenflywheel commented on src/test/java/emissary/test/core/junit5/ExtractionTest.java:
```diff
@@ -171,8 +171,12 @@ protected void checkAnswers(Document answers, IBaseDataObject payload, List<IBas
         int numAtt = JDOMUtil.getChildIntValue(el, "numAttachments");
+        long numAttElements = el.getChildren().stream().filter(c -> c.getName().startsWith("att")).count();
+        if (Math.toIntExact(numAttElements) != numAtt && !(numAttElements <= 0 && numAtt == -1)) {
+            logger.warn("Expected numAttachments and actual <att#> count in {} not equal. Expected: {} Actual: {}", tname, Math.max(numAtt, 0),
```
We have plenty of downstream answer files that only spot check among numerous attachments.  I should be worried if the number of `<attN>` elements exceeds the `<numAttachments>` value, but I'd hate to add hundreds of `<attN>` elements just to suppress this warning for a given test.  

Maybe we can add another element that allows "attN count validation" behavior to be overridden for the current answer file.  Something like `<attCountCheck>minimal</attCountCheck>` ?   (Naming things is hard.)

sambish5 replied:
I think that is something that could be added if that was the desire. I don't know the best way to go about it/if we'd want different logging levels if counts were excessively different, but the basic idea I agree could be helpful in those downstream environments.

drivenflywheel replied:
Another option (maybe better than the new answer file element) would be to add a private member to ExtractionTest that governs this behavior.  If desired, extending test classes could disable/alter the behavior by calling a setting that modifies the value of that private member -  e.g., `setAttachmentCountValidation(newValue);`

sambish5 replied:
That I think may end up being a cleaner/easier option in the long run. Would we only want to do this for attachments, or extracts as well?

drivenflywheel commented on src/test/java/emissary/test/core/junit5/ExtractionTest.java:
```diff
@@ -419,4 +433,13 @@ protected void setupPayload(IBaseDataObject payload, Document doc) {
     }
+
+    // allow the validation of att and ext counts to be skipped/not logged in tests
+    protected void setAttachmentCountValidation(boolean skipAttCountValidation) {
+        this.skipAttCountValidation = skipAttCountValidation;
+    }
+
+    protected void setExtractCountValidation(boolean skipExtractCountValidation) {
+        this.skipExtractCountValidation = skipExtractCountValidation;
+    }
```
These parameter names are unexpected, since they're the inversion of what the method name would suggest.  Should rename to clarify
```suggestion
    protected void setAttachmentCountValidation(boolean value) {
        this.skipAttCountValidation = !value;
    }

    protected void setExtractCountValidation(boolean value) {
        this.skipExtractCountValidation = !value;
    }
```

drivenflywheel commented on src/test/java/emissary/test/core/junit5/ExtractionTest.java:
```diff
@@ -171,8 +174,12 @@ protected void checkAnswers(Document answers, IBaseDataObject payload, List<IBas
         int numAtt = JDOMUtil.getChildIntValue(el, "numAttachments");
+        long numAttElements = el.getChildren().stream().filter(c -> c.getName().startsWith("att")).count();
+        if (Math.toIntExact(numAttElements) != numAtt && !(numAttElements <= 0 && numAtt == -1) && !skipAttCountValidation) {
+            logger.warn("Expected numAttachments and actual <att#> count in {} not equal. Expected: {} Actual: {}", tname, Math.max(numAtt, 0),
+                    numAttElements);
```
Now that we've provided a way to disable validation, we should make this an assertion by default.

```suggestion
            if (numAtt >= -1 || numAttElements > 0) {
                if (!skipAttCountValidation) {
                    assertEquals(numAtt, Math.toIntExact(numAttElements), "Expected numAttachments not equal to number of <att#> elements");
                } else {
                    if (Math.toIntExact(numAttElements) != numAtt) {
                        logger.warn("Expected numAttachments and actual <att#> count in {} not equal. Expected: {} Actual: {}", tname,
                            Math.max(numAtt, 0), numAttElements);
                    }
                }
            }
```

sambish5 replied:
so you still want it to log the mismatched counts even with validation skipped? From the previous discussion, it seemed like we wanted the logging to only happen when validation was desired, so that in files where the numbers were super far off, it could be disabled and the warning would not constantly be yelled about.

drivenflywheel reviewed (changes requested):
Last changes? 🤞

sambish5 commented:
@drivenflywheel made the suggested changes, had to make some slight changes to the assertion things to make it work  like how it did when it was just logging

drivenflywheel reviewed (dismissed):
Looks good.  Sorry for the high number cycles.

sambish5 commented:
rebased and made some _minor_ updates to just include the test name when validation finds expected vs found count not equal on assertion, hopefully this can close this PR 🤞

drivenflywheel commented on src/test/java/emissary/test/core/junit5/ExtractionTest.java:
```diff
@@ -280,15 +295,29 @@ protected void checkAnswers(Element el, IBaseDataObject payload, @Nullable List<
-                assertEquals(Integer.parseInt(extractCountStr), foundCount,
-                        String.format("Number of extracted children in '%s' is %s, not expected %s", tname, foundCount, extractCountStr));
+            if (extractCount > -1) {
+                assertEquals(extractCount, foundCount,
```
Should this get the same `skipExtractCountValidation` treatment as above? (warn vs assert)

sambish5 replied:
I could, though that is kind of out of the scope of this PR. The check above this verifies answer files have matching counts, I'm not sure if we want to have the option to "skip validation" when actually checking the answerFile vs the explicit payload extracted records count

drivenflywheel replied:
Ah - you're 100% right.  I lost track of which count the various variables were referencing.

sambish5 commented:
Changes made by request of @jpdahlke 
Overview of changes:
- no longer fails if `<numAttachments>` & `<extractCount>` DNE `<att#>` or `<extract#>` in answer file, now just logs differences.
- Has a new priority system to make sure that counts in answer file match counts in payload:
       - Checks `<numAttachments>`/`<extractCount>` if present in answer file against count from payload
       - If those are not present, check against `<att#>`/`<extract#>` if present
       - If neither are present, and there are atts or extracts in the payload, **FAIL** the test, saying `<numAttachments>`/`<extractCount>` matching payload count needs to be in answer file

jpdahlke commented on src/test/java/emissary/test/core/junit5/ExtractionTest.java:
```diff
@@ -173,8 +173,26 @@ protected void checkAnswers(Element el, IBaseDataObject payload, @Nullable List<
+            if (Math.toIntExact(numAttElements) != numAtt) {
+                logger.warn("Expected and actual attachment counts in {} not equal. <numAttachments>: {} <att#>: {}", tname,
+                        Math.max(numAtt, 0), numAttElements);
+            }
```
I would just remove these log messages I think. Based on the behavior above we expect this scenario and accept it.

sambish5 replied:
You're right, I'll remove them

jpdahlke reviewed (commented):
I think this update makes sense given our integration experiences.
